### PR TITLE
[clang] Change representation of CurLexerKind

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Previous representation used an enumeration combined to a switch to dispatch to the appropriate lexer.

Use function pointer so that the dispatching is just an indirect call, which is actually better because lexing is a costly task compared to a function call.

This also makes the code slightly cleaner, speedup on compile time tracker are consistent and range form -0.05% to -0.20% for NewPM-O0-g, see

        https://llvm-compile-time-tracker.com/compare.php?from=f9906508bc4f05d3950e2219b4c56f6c078a61ef&to=608c85ec1283638db949d73e062bcc3355001ce4&stat=instructions:u

Considering just the preprocessing task, preprocessing the sqlite amalgametion takes -0.6% instructions (according to valgrind --tool=callgrind)

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
Updated compile time metrics: https://llvm-compile-time-tracker.com/compare.php?from=f9906508bc4f05d3950e2219b4c56f6c078a61ef&to=2631a5163de552a8e482f5cbc51525fb8bb5082f&stat=instructions:u

File: clang/lib/Lex/Preprocessor.cpp
Comment (cor3ntin) [inline clang/lib/Lex/Preprocessor.cpp:646]: I don't think you need the cast to void here

**Review by cor3ntin — COMMENTED** [review]
This is excellent!
I made some naming suggestion. Now that we don't have an enum i think it is clearer to drop the "kind" terminology

**Comment by serge-sans-paille** [comment]
Agreed. Sorry I pushed --force my change out of habit :-/

File: clang/utils/ClangVisualizers/clang.natvis
Comment (cor3ntin) [inline clang/utils/ClangVisualizers/clang.natvis:820]: @AaronBallman can you review this change?

File: clang/include/clang/Lex/Preprocessor.h
Comment (cor3ntin) [inline clang/include/clang/Lex/Preprocessor.h:1912]: I think renaming this one is a bit too aggressive. Ideally that we use a callback is an implementation detail that should not leak more than it needs to

 ```suggestion
  void recomputeCurLexerKind();
```

File: clang/include/clang/Lex/Preprocessor.h
Comment (AaronBallman) [inline clang/include/clang/Lex/Preprocessor.h:765]: Can you align this with the other nearby declarations?

File: clang/utils/ClangVisualizers/clang.natvis
Comment (AaronBallman) [reply clang/utils/ClangVisualizers/clang.natvis:820]: Yeah, this is going to break NATVIS pretty badly. This whole nonsense:
```
IncludeMacroStack._Mypair._Myval2._Mylast - IncludeMacroStack._Mypair._Myval2._Myfirst
```
is a ridiculous^Wfancy way of saying:
```
IncludeMacroStack.back().CurLexerKind
```
which is used to figure out what kind of lexer we should be displaying a visualizer for. I don't think there's a way we can repair that now that we're using function pointers. I checked out the behavior locally to see if the debugging experience is significantly worse and... it's not *that* bad (I think the existing NATVIS already needed some repair). The most confusing thing is that we no longer can quickly see which lexer is the one currently in use, but we can still access the correct lexer from other contextual information.

So I think this change is "fine".

File: clang/include/clang/Lex/Preprocessor.h
Comment (serge-sans-paille) [reply clang/include/clang/Lex/Preprocessor.h:765]: That would break clang-format :-) If we want to keep the original alignement, I need to add clang-format comments, should I do it that way?

Comment (AaronBallman) [reply clang/include/clang/Lex/Preprocessor.h:765]: Go ahead and break clang-format; it's not paying attention to the surrounding coding style which is what we're supposed to do per community policy. No need to add clutter comments.

**Comment by serge-sans-paille** [comment]
As expected, the formatter iw now red, otherwise looks good (to me at least), waiting for official approval :-)

**Review by AaronBallman — APPROVED** [review]
LGTM! It's always great to improve compile-time performance!

**Pull request merged**
